apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4SamhsaMatcher.java

```diff
@@ -522,43 +524,79 @@ private static String normalizeIcd10Code(String icd10DiagnosisCode) {
+       * SAMHSA.
+       */
+      return true;
+    } else {
+      // Otherwise, this only has known & non-SAMHSA-blacklisted procedure codes.
+      return false;
+    }
   }
 
   /**
-   * @param procedureCoding the procedure {@link Coding} to check
-   * @return <code>true</code> if the specified procedure {@link Coding} matches one of the {@link
-   *     #cptCodes} entries, <code>false</code> if it does not
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains at least
+   *     one {@link Coding} and only contains {@link Coding}s that have known coding systems <code>
+   *     false</code> otherwise
+   */
+  private boolean containsOnlyKnownSystems(CodeableConcept procedureConcept) {
+    Set<String> codingSystems =
+        procedureConcept.getCoding().stream().map(Coding::getSystem).collect(Collectors.toSet());
+
+    String hcpcsCdSystem = CCWUtils.calculateVariableReferenceUrl(CcwCodebookVariable.HCPCS_CD);
+
+    // Valid system url for productOrService coding
+    Set<String> hcpcsSystem = Set.of(TransformerConstants.CODING_SYSTEM_HCPCS);
+
+    // Additional valid coding system URL for backwards-compatibility
+    // See: https://jira.cms.gov/browse/BFD-1345
+    Set<String> backwardsCompatibleHcpcsSystems =
+        Set.of(TransformerConstants.CODING_SYSTEM_HCPCS, hcpcsCdSystem);
+
+    return codingSystems.equals(hcpcsSystem)
+        || codingSystems.equals(backwardsCompatibleHcpcsSystems);
+  }
+
+  /**
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains any
+   *     {@link Coding}s that match any of the {@link #cptCodes}, <code>false</code> if they all do
+   *     not
    */
-  private boolean isSamhsaCptCode(Coding procedureCoding) {
+  private boolean hasHspcAndSamhsaCptCode(CodeableConcept procedureConcept) {
     /*
      * Note: CPT codes represent a subset of possible HCPCS codes (but are the only
      * subset that we blacklist from).
      */
-    if (!TransformerConstants.CODING_SYSTEM_HCPCS.equals(procedureCoding.getSystem()))
-      throw new IllegalArgumentException();
+    Set<String> codingSystems =
+        procedureConcept.getCoding().stream().map(Coding::getSystem).collect(Collectors.toSet());
+
+    // Does the CodeableConcept have a legit HCPCS Coding?
+    if (!codingSystems.contains(TransformerConstants.CODING_SYSTEM_HCPCS)) {
+      return false;
+    }
 
     /*
      * Note: per XXX all codes in icd10DiagnosisCodes are already normalized.
      */
-    return cptCodes.contains(normalizeHcpcsCode(procedureCoding.getCode()));
+    return procedureConcept.getCoding().stream()
+        .anyMatch(
+            procedureCoding -> cptCodes.contains(normalizeHcpcsCode(procedureCoding.getCode())));
```

Comment:
I was noticing this seems different from the existing checks.  `isSamhsaCptCode` and `isSamhsaIcd9[10]procedure[diagnosis]` all throw if it's the wrong system, where this (`hasHspcAndSamhsaCptCode`) just returns false.  Just wondering if it would be good to make all the API methods a little more consistent in behavior?

Reply:
Could use some domain knowledge input here, do we expect HCPCS to _always_ exist? Theres a bit of a difference between gracefully bowing out of the filter check with false, and blowing up with an exception. (Maybe the other methods could/should be changed to be more graceful, or maybe this one should be more restrictive, depending on the answer to that).

Reply:
Thanks for the insight Karl! I'll adjust this to be an exception (and fix the tests to expect that for a wrong system scenario).

Reply:
One thing to note about this though, the same test for an unknown system does NOT cause an exception in v1. Returning false actually matches v1 functionality, for consistency. Maybe that means v1 is less secure and we've "upgraded" v2, but just an FYI.

Reply:
I did not explain this particularly well.

@lsmitchell Is this method ever called for `CodeableConcept`s that don't contain the HCPCS code? Like, on purpose? If so, throwing an exception is inappropriate. If not, it is.

Reply:
Since DME can have other codes, switched back to false and left a comment noting why.